torch/utils/_device.py

```diff
@@ -65,12 +66,36 @@ def __enter__(self):
         global CURRENT_DEVICE
         self.old_device = CURRENT_DEVICE
         CURRENT_DEVICE = self.device
-        return super().__enter__()
+        # We need to put the device at the bottom of the stack
```

Comment:
hmm cc @albanD do you mind reviewing this too?

In particular - I'm not sure which implementation we want, between:

(1) this one (invariant that the `DeviceContext` mode only ever appears once on the stack)

(2) treat the mode more like an "ordinary" mode: you can push more than one of them on the stack, and the top-most mode will always take precedence. E.g. if you push `context1` onto the stack, then `context2`, `context2` will shadow the `context1`. when `context2` is popped later, `context1` applies.

Maybe related question for @mlazos - do you think (2) would make it easier to support this mode in dynamo without extra special handling?

Reply:
Talked offline - it seems like given the fact that `set_default_device(None)` is supposed to pop the stack, we fundamentally cannot treat this mode as an "ordinary" mode (it's push/pop are not guaranteed to happen in the same order compared to other modes)